File: skills/go-scaffold/scripts/templates/files/internal/shared/auth/rbac.go
package auth

// Static role→permission map (ADR §7). Deliberately NOT a DB table for a
// two-module scaffold — a documented extension point, not a gap. To move it to
// the DB, replace rolePermissions with a lookup and keep Can()'s signature.
//
// Can() is called INSIDE application/ use-cases, never in api/ handlers: the
// auth middleware only proves identity; authorization is a separate, explicit
// per-use-case check.
type Role = string
type Permission = string

const (
	RoleAdmin Role = "admin"
	RoleUser  Role = "user"

	PermUsersRead  Permission = "users:read"
	PermUsersErase Permission = "users:erase"
	PermPostsRead  Permission = "posts:read"
	PermPostsWrite Permission = "posts:write"
)

var rolePermissions = map[Role][]Permission{
	RoleAdmin: {PermUsersRead, PermUsersErase, PermPostsRead, PermPostsWrite},
	RoleUser:  {PermUsersRead, PermPostsRead, PermPostsWrite},
}

// Can reports whether any of the caller's roles grants the permission. Unknown
// roles are silently ignored.
func Can(roles []string, permission Permission) bool {
	for _, r := range roles {
		for _, p := range rolePermissions[r] {
			if p == permission {
				return true
			}
		}
	}
	return false
}

File: skills/nodejs-scaffold/scripts/templates/files/src/shared/auth/rbac.ts
// Static role→permission map. Deliberately NOT a DB table for a two-module
// scaffold — a documented extension point, not a gap. To move it to the DB,
// replace ROLE_PERMISSIONS with a lookup and keep the `can()` signature.
//
// `can()` is called INSIDE application/ use-cases, never in api/ handlers:
// app.authenticate only proves identity; authorization is a separate, explicit
// per-use-case check.
export type Role = 'admin' | 'user'
export type Permission = 'users:read' | 'users:write' | 'users:erase' | 'posts:read' | 'posts:write'

export const ROLE_PERMISSIONS: Record<Role, readonly Permission[]> = {
  admin: ['users:read', 'users:write', 'users:erase', 'posts:read', 'posts:write'],
  user: ['users:read', 'posts:read', 'posts:write']
}

function isRole(value: string): value is Role {
  return value === 'admin' || value === 'user'
}

export function can(roles: string[], permission: Permission): boolean {
  return roles.some((r) => isRole(r) && ROLE_PERMISSIONS[r].includes(permission))
}
